crates/unladen_swallow/src/wcc.rs

```diff
@@ -0,0 +1,67 @@
+use crate::graphs::{NumpyType, SharedSlice};
+use graph::prelude::{
+    wcc_afforest as graph_wcc, Components, DirectedDegrees, DirectedNeighbors, Graph as GraphTrait,
+    Idx, WccConfig,
+};
+use numpy::PyArray1;
+use pyo3::prelude::*;
+use std::hash::Hash;
+use std::time::{Duration, Instant};
+
+pub(crate) fn register(_py: Python<'_>, m: &PyModule) -> PyResult<()> {
+    m.add_class::<WccResult>()?;
+    Ok(())
+}
+
+pub(crate) fn wcc<NI, G, C>(py: Python<'_>, graph: &G, config: C) -> WccResult
+where
+    NI: Idx + Hash + NumpyType,
+    G: GraphTrait<NI> + DirectedDegrees<NI> + DirectedNeighbors<NI> + Sync,
+    C: Into<Option<WccConfig>> + Send,
+{
+    py.allow_threads(move || inner_wcc(graph, config))
+}
+
+fn inner_wcc<NI, G>(graph: &G, config: impl Into<Option<WccConfig>>) -> WccResult
+where
+    NI: Idx + Hash + NumpyType,
+    G: GraphTrait<NI> + DirectedDegrees<NI> + DirectedNeighbors<NI> + Sync,
+{
+    let config = config.into().unwrap_or_default();
+    let start = Instant::now();
+    let components = graph_wcc(graph, config).to_vec();
+    let micros = start.elapsed().as_micros().min(u64::MAX as _) as _;
+    let components = SharedSlice::from_vec(components);
+    WccResult { components, micros }
+}
+
+#[pyclass]
+#[derive(Clone)]
+pub struct WccResult {
+    components: SharedSlice,
+    #[pyo3(get)]
+    micros: u64,
+}
+
+impl std::fmt::Debug for WccResult {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("WccResult")
+            .field(
+                "components",
+                &format!("[... {} values]", self.components.len()),
+            )
+            .field("took", &Duration::from_micros(self.micros))
+            .finish()
+    }
+}
+
+#[pymethods]
+impl WccResult {
+    pub fn components<'py>(&self, py: Python<'py>) -> PyResult<&'py PyArray1<u32>> {
```

Comment:
@knutwalker Following your page rank example, exposing wcc was straightforward :+1:

There is one problem, however. Wcc (and other algos, too) use the given `NI: Idx` type in their result vecs. Since we cannot make `pymethods` generic, I'm currently using a fixed type. One thing we could do it converting them to `usize` via `NI::index` but I assume this won't help as it can still be `u32` or `u64`. Any ideas?